### PR TITLE
[EVM] Refactor EVMControlFlowGraph/EVMControlFlowGraphBuilder

### PR DESCRIPTION
**Pull request by PavelKopyl**
- Removed EVMControlFlowGraph/EVMControlFlowGraphBuilder classes
- Functionality that analyzes machine CFG and provides query methods was moved to EVMMachineCFGInfo class
- Information about MBB terminators is represented in EVMMBBTerminatorsInfo class
- StackSlot, Stack and Operation definitions  were moved to EVMStackModel
- Replaced almost all the std::map/std::set with llvm counterparts in EVMStackLayoutGenerator

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      UnconditionsBranch = &*I;
+    else if (I->isConditionalBranch())
+      ConditionaBranch = &*I;
+    ++I;
+  }
+  return std::tie(ConditionaBranch, UnconditionsBranch);
+}
+
+static bool isFunctionReturn(const MachineInstr *MI) {
+  return MI->getOpcode() == EVM::RET;
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: I don't think we should mark `RETURN`, `REVERT` with `isReturn`. EraVM doesn't. So we can use `isReturn` tablegenned flag.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+static bool isTerminate(const MachineInstr *MI) {
+  unsigned Opc = MI->getOpcode();
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: nit: unnecessary temporary.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+#include "EVMMachineCFGInfo.h"
+#include "EVMHelperUtilities.h"
+#include "EVMMachineFunctionInfo.h"
+#include "EVMSubtarget.h"
+#include "MCTargetDesc/EVMMCTargetDesc.h"
+#include "llvm/CodeGen/MachineFunction.h"
+
+using namespace llvm;
+
+static std::tuple<MachineInstr *, MachineInstr *>
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: We can use pair here, instead of tuple.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+  MachineInstr *UnconditionsBranch = nullptr;
+  MachineBasicBlock::reverse_iterator I = MBB.rbegin(), E = MBB.rend();
+  while (I != E && I->isTerminator()) {
+    if (I->isUnconditionalBranch())
+      UnconditionsBranch = &*I;
+    else if (I->isConditionalBranch())
+      ConditionaBranch = &*I;
+    ++I;
+  }
+  return std::tie(ConditionaBranch, UnconditionsBranch);
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]:
```suggestion
  return {ConditionaBranch, UnconditionsBranch};
```

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  return MI->getOpcode() == EVM::RET;
+}
+
+static bool isTerminate(const MachineInstr *MI) {
+  unsigned Opc = MI->getOpcode();
+  switch (Opc) {
+  default:
+    return false;
+  case EVM::REVERT:
+    [[fallthrough]];
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: No need for using `[[fallthrough]]` in this switch.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  // or a 'noreturn' function is at the end of MBB.
+  if (!TBB && !FBB && MBB.succ_empty()) {
+    Info->ExitType = MBBExitType::Terminate;
+    return;
+  }
+
+  bool IsLatch = false;
+  if (MachineLoop *ML = MLI->getLoopFor(&MBB)) {
+    SmallVector<MachineBasicBlock *, 8> Latches;
+    ML->getLoopLatches(Latches);
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: Can we use `isLoopLatch` instead of traversing over all latches?

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+  Info->TrueBB = TBB;
+  Info->FalseBB = FBB;
+  Info->IsFallThrough = IsFallThrough;
+  Info->CondBr = CondBr;
+  Info->UncondBr = UncondBr;
+}
+
+// Mark basic blocks that have outgoing paths to function returns.
+void EVMMachineCFGInfo::collectBlocksLeadingToFunctionReturn(
+    const MachineBasicBlock *Return) {
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: In case we have multiple function returns, we could end up traversing the same path to the entry more than once. I suggest to pass all return MBBs, and to use worklist algorithm so we don't traverse the same path more than once.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.h [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+enum class MBBExitType {
+  Invalid,
+  ConditionalBranch,
+  UnconditionalBranch,
+  FunctionReturn,
+  Terminate
+};
+
+class EVMMBBTerminatorsInfo {
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.h:38]: Maybe we can group data members in union, something similar to what is implemented in `MachineOperand`? That way we will reduce memory + code will be self explanatory based on the enum above.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.h [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  friend class EVMMachineCFGInfo;
+
+  MBBExitType ExitType = MBBExitType::Invalid;
+  SmallVector<MachineOperand, 1> Condition;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.h]: Do we need a vector here?

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.h [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  MachineBasicBlock *TrueBB = nullptr;
+  MachineBasicBlock *FalseBB = nullptr;
+  MachineInstr *CondBr = nullptr;
+  MachineInstr *UncondBr = nullptr;
+  MachineInstr *LastTerm = nullptr;
+  bool IsFallThrough = false;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.h]: As far as I can see this data member is only used for debug prints, so I think we can remove it.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+                                     MachineLoopInfo *MLI)
+    : MF(MF), MLI(MLI) {
+  for (MachineBasicBlock &MBB : MF)
+    collectTerminatorsInfo(MBB);
+
+  for (const MachineBasicBlock &MBB : MF) {
+    const EVMMBBTerminatorsInfo *TermInfo = getTerminatorsInfo(&MBB);
+    if (TermInfo->getExitType() == MBBExitType::FunctionReturn)
+      collectBlocksLeadingToFunctionReturn(&MBB);
+    collectCutVertexes(&MBB);
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: Do we need to call this function for every MBB or only for entry MBB?

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.h [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  bool isCutVertex(const MachineBasicBlock *MBB) const {
+    return CutVertexes.count(MBB) > 0;
+  }
+
+  bool isOnPathToFuncReturn(const MachineBasicBlock *MBB) const {
+    return ToFuncReturnVertexes.count(MBB) > 0;
+  }
+
+private:
+  MachineFunction &MF;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.h]: Do we need this data member?

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.h [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  MachineLoopInfo *MLI = nullptr;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.h]: Do we need this data member?

File: llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp [hunk 14]
```diff
@@ -447,36 +438,33 @@ void EVMStackifyCodeEmitter::createOperationLayout(const CFG::Operation &Op) {
   assert(areLayoutsCompatible(StackInput, Op.Input));
 }
 
-void EVMStackifyCodeEmitter::run(CFG::BasicBlock &EntryBB) {
+void EVMStackifyCodeEmitter::run(MachineBasicBlock *EntryBB) {
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp]: We can remove EntryBB argument and use `MF.front()` to initialize `WorkList`.

**Review by vladimirradosavljevic — COMMENTED** [review]
Overall, the code looks good. I just have a few comments.

**Comment by PavelKopyl** [comment]
Hey @vladimirradosavljevic, thank you for the review. I agree with all the notes. Please find the fixes in the last commit.

**Review by akiramenai — APPROVED** [review]
LGTM, thanks

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+    if (!TBB) {
+      assert(!UncondBr);
+      assert(MBB.getSingleSuccessor());
+      TBB = MBB.getFallThrough();
+      assert(TBB);
+    }
+    Info->ExitType = MBBExitType::UnconditionalBranch;
+    Info->BranchInfo.Unconditional.TargetBB = TBB;
+    Info->BranchInfo.Unconditional.Br = UncondBr;
+    Info->BranchInfo.Unconditional.IsLatch = IsLatch;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]:
```suggestion
    Info->BranchInfo.Unconditional = {TBB, UncondBr, IsLatch};
```

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+      FBB = MBB.getFallThrough();
+      assert(FBB);
+    }
+    Info->ExitType = MBBExitType::ConditionalBranch;
+    Info->BranchInfo.Conditional.TrueBB = TBB;
+    Info->BranchInfo.Conditional.FalseBB = FBB;
+    Info->BranchInfo.Conditional.CondBr = CondBr;
+    Info->BranchInfo.Conditional.UncondBr = UncondBr;
+    assert(Cond[0].isIdenticalTo(CondBr->getOperand(1)));
+    Info->BranchInfo.Conditional.Condition = &CondBr->getOperand(1);
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]:
```suggestion
    assert(Cond[0].isIdenticalTo(CondBr->getOperand(1)));
    Info->BranchInfo.Conditional = {&CondBr->getOperand(1), TBB, FBB, CondBr,
                                    UncondBr};
```

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+#include "MCTargetDesc/EVMMCTargetDesc.h"
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+
+using namespace llvm;
+
+static std::pair<MachineInstr *, MachineInstr *>
+getBranchInstructions(MachineBasicBlock &MBB) {
+  MachineInstr *ConditionaBranch = nullptr;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: Typo: `ConditionalBranch`

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+  MachineInstr *UnconditionsBranch = nullptr;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: Typo: `UnconditionalBranch`

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+    assert(Cond[0].isIdenticalTo(CondBr->getOperand(1)));
+    Info->BranchInfo.Conditional.Condition = &CondBr->getOperand(1);
+  }
+}
+
+// Mark basic blocks that have outgoing paths to function returns.
+void EVMMachineCFGInfo::collectBlocksLeadingToFunctionReturn(
+    SmallVector<const MachineBasicBlock *> &Returns) {
+  SmallPtrSet<const MachineBasicBlock *, 32> Visited;
+  for (const MachineBasicBlock *RetMBB : Returns) {
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: I think we can initialize `WorkList` with `Returns` and remove this for loop.

File: llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp [hunk 20]
```diff
@@ -489,83 +476,63 @@ void EVMStackifyCodeEmitter::run(CFG::BasicBlock &EntryBB) {
...
+
+      if (UncondBr)
+        Emitter.emitUncondJump(UncondBr, Target);
+      WorkList.emplace_back(Target);
+    } else if (ExitType == MBBExitType::ConditionalBranch) {
+      MachineBasicBlock *TrueBB = nullptr, *FalseBB = nullptr;
+      MachineInstr *CondBr = nullptr, *UncondBr = nullptr;
+      const MachineOperand *Condition = nullptr;
+      std::tie(CondBr, UncondBr, TrueBB, FalseBB, Condition) =
+          TermInfo->getConditionalBranch();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp]:
```suggestion
      auto [CondBr, UncondBr, TrueBB, FalseBB, Condition] =
          TermInfo->getConditionalBranch();
```

File: llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp [hunk 21]
```diff
@@ -489,83 +476,63 @@ void EVMStackifyCodeEmitter::run(CFG::BasicBlock &EntryBB) {
...
-                llvm_unreachable("Unexpected BB terminator");
-            }},
-        Block->Exit);
+    const EVMMBBTerminatorsInfo *TermInfo = CFGInfo.getTerminatorsInfo(Block);
+    MBBExitType ExitType = TermInfo->getExitType();
+    if (ExitType == MBBExitType::UnconditionalBranch) {
+      MachineBasicBlock *Target = nullptr;
+      MachineInstr *UncondBr = nullptr;
+      std::tie(UncondBr, Target, std::ignore) =
+          TermInfo->getUnconditionalBranch();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp]:
```suggestion
      auto [UncondBr, Target, _] = TermInfo->getUnconditionalBranch();
```

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.h [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  std::tuple<MachineInstr *, MachineInstr *, MachineBasicBlock *,
+             MachineBasicBlock *, const MachineOperand *>
+  getConditionalBranch() const {
+    assert(ExitType == MBBExitType::ConditionalBranch);
+    return std::make_tuple(
+        BranchInfo.Conditional.CondBr, BranchInfo.Conditional.UncondBr,
+        BranchInfo.Conditional.TrueBB, BranchInfo.Conditional.FalseBB,
+        BranchInfo.Conditional.Condition);
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.h]:
```suggestion
    return {BranchInfo.Conditional.CondBr, BranchInfo.Conditional.UncondBr,
            BranchInfo.Conditional.TrueBB, BranchInfo.Conditional.FalseBB,
            BranchInfo.Conditional.Condition};
```

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.h [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  std::tuple<MachineInstr *, MachineBasicBlock *, bool>
+  getUnconditionalBranch() const {
+    assert(ExitType == MBBExitType::UnconditionalBranch);
+    return std::make_tuple(BranchInfo.Unconditional.Br,
+                           BranchInfo.Unconditional.TargetBB,
+                           BranchInfo.Unconditional.IsLatch);
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMMachineCFGInfo.h]:
```suggestion
    return {BranchInfo.Unconditional.Br, BranchInfo.Unconditional.TargetBB,
            BranchInfo.Unconditional.IsLatch};
```

File: llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp [hunk 24]
```diff
@@ -398,183 +390,163 @@ void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const &Entry) {
...
+std::optional<Stack> EVMStackLayoutGenerator::getExitLayoutOrStageDependencies(
+    const MachineBasicBlock *Block,
+    const DenseSet<const MachineBasicBlock *> &Visited,
+    std::list<const MachineBasicBlock *> &ToVisit) const {
+  const EVMMBBTerminatorsInfo *TermInfo = CFGInfo.getTerminatorsInfo(Block);
+  MBBExitType ExitType = TermInfo->getExitType();
+  if (ExitType == MBBExitType::UnconditionalBranch) {
+    const MachineBasicBlock *Target = nullptr;
+    bool IsLatch = false;
+    std::tie(std::ignore, Target, IsLatch) = TermInfo->getUnconditionalBranch();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp]:
```suggestion
    auto [_, Target, IsLatch] = TermInfo->getUnconditionalBranch();
```

File: llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp [hunk 25]
```diff
@@ -398,183 +390,163 @@ void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const &Entry) {
...
+    // Otherwise stage the jump target for visit and defer the current
+    // block.
+    ToVisit.emplace_front(Target);
+    return std::nullopt;
+  }
+  if (ExitType == MBBExitType::ConditionalBranch) {
+    const MachineBasicBlock *TrueBB = nullptr, *FalseBB = nullptr;
+    const MachineOperand *Condition = nullptr;
+    std::tie(std::ignore, std::ignore, TrueBB, FalseBB, Condition) =
+        TermInfo->getConditionalBranch();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp]:
```suggestion
    auto [CondBr, UncondBr, TrueBB, FalseBB, Condition] =
        TermInfo->getConditionalBranch();
```
Not sure this is better, since `std::ignore` can't be used in structured bindings.

File: llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp [hunk 26]
```diff
@@ -398,183 +390,163 @@ void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const &Entry) {
...
+  EVMUtils::BreadthFirstSearch<MachineBasicBlock const *>{{Entry}}.run(
+      [&](const MachineBasicBlock *Block, auto VisitSucc) {
+        const EVMMBBTerminatorsInfo *TermInfo =
+            CFGInfo.getTerminatorsInfo(Block);
+        MBBExitType ExitType = TermInfo->getExitType();
+        if (ExitType == MBBExitType::UnconditionalBranch) {
+          MachineBasicBlock *Target = nullptr;
+          bool IsLatch = false;
+          std::tie(std::ignore, Target, IsLatch) =
+              TermInfo->getUnconditionalBranch();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp]:
```suggestion
          auto [_, Target, IsLatch] = TermInfo->getUnconditionalBranch();
```

File: llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp [hunk 27]
```diff
@@ -398,183 +390,163 @@ void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const &Entry) {
...
+    const MachineBasicBlock *Block) {
+  EVMUtils::BreadthFirstSearch<MachineBasicBlock const *> BFS{{Block}};
+  BFS.run([&](const MachineBasicBlock *Block, auto VisitSucc) {
+    const EVMMBBTerminatorsInfo *TermInfo = CFGInfo.getTerminatorsInfo(Block);
+    MBBExitType ExitType = TermInfo->getExitType();
+    if (ExitType == MBBExitType::UnconditionalBranch) {
+      MachineBasicBlock *Target = nullptr;
+      bool IsLatch = false;
+      std::tie(std::ignore, Target, IsLatch) =
+          TermInfo->getUnconditionalBranch();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp]:
```suggestion
      auto [_, Target, IsLatch] = TermInfo->getUnconditionalBranch();
```

File: llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp [hunk 28]
```diff
@@ -398,183 +390,163 @@ void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const &Entry) {
...
+      if (!IsLatch)
+        VisitSucc(Target);
+      return;
+    }
+    if (ExitType == MBBExitType::ConditionalBranch) {
+      MachineBasicBlock *TrueBB = nullptr, *FalseBB = nullptr;
+      const MachineOperand *Condition = nullptr;
+      std::tie(std::ignore, std::ignore, TrueBB, FalseBB, Condition) =
+          TermInfo->getConditionalBranch();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp]:
```suggestion
      auto [CondBr, UncondBr, TrueBB, FalseBB, Condition] =
          TermInfo->getConditionalBranch();
```
Not sure this is better, since `std::ignore` can't be used in structured bindings.

**Review by vladimirradosavljevic — APPROVED** [review]
LGTM, thanks! Just small nitpicking.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: Hm.. as I see we use isReturn in EraVM,
```
class IRetBase<EraVMOpcode opcode,
               list<EraVMOperand> ops,                                  
               list<dag> pattern>
  : IForm <opcode, ops, pattern> {                                              

  let isReturn = 1;                                                             
  let isTerminator = 1;                    
  let isBarrier = 1;                                                            
}
```

Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: But anyway I removed the flag.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: Thanks.

File: llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp (hunk 19 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMMachineCFGInfo.cpp]: Thanks!

File: llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp (hunk 25 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMStackLayoutGenerator.cpp]: I like the structural bindings here, as it reduces the code. The main reason I not used is to avoid possible 'unused variable'. But OK, it seems the '_' works as  name-independent in our build environment.

**Comment by akiramenai** [comment]
Please squash and merge
